File: src/6_printListFromTailToHead.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <cassert>
#include <tuple>
#include <memory>

#include "offerhelper.h"

using namespace std;

class BaseSolution
{
public:
    virtual std::vector<int> printListFromTailToHead(ListNode* head) = 0;
};


/**
 * Simple use a stack.
 */
class Solution: public virtual BaseSolution
{
public:
    vector<int> printListFromTailToHead(ListNode* head) override
    {
        std::vector<int> vec;
        while (head)
        {
            vec.push_back(head->val);
            head = head->next;
        }
        reverse(vec.begin(), vec.end());
        return vec;
    }
};


/**
 * Insert at the begining.
 */
class Solution1: public virtual BaseSolution
{
public:
    vector<int> printListFromTailToHead(ListNode* head) override
    {
        // insertion
        ListNode *newHead = new ListNode(0);
        while (head)
        {
            ListNode *saved_next = head->next;
            head->next = newHead->next;
            newHead->next = head;
            head = saved_next;
        }

        std::vector<int> vec;
        head = newHead->next;
        while (head)
        {
            vec.push_back(head->val);
            head = head->next;
        }
        return vec;
    }
};

/**
 * Recursion soluton.
 */
class Solution2: public virtual BaseSolution
{
public:
    vector<int> printListFromTailToHead(ListNode* head) override
    {
        std::vector<int> vec;
        if (head)
        {
            auto ret = printListFromTailToHead(head->next);
            vec.insert(vec.end(), ret.begin(), ret.end());
            vec.push_back(head->val);
        }
        return vec;
    }
};

/**
 * Use either move semantic or reference will be fine.
 */
void test(unique_ptr<BaseSolution> pSolution)
{
    int n = 10000;
    std::vector<int> largeRandomVector = Offer::genRandomVector(n, 0, n);
    std::vector<vector<int>> testCases{
        {},
        {1},
        {1, 2, 3},
        largeRandomVector,
    };

    for (auto v : testCases)
    {
        // cout << "testcase: ";
        // Offer::printVec(v);
        ListNode* head = Offer::createLinkedList(v);
        auto ret = pSolution->printListFromTailToHead(head);
        reverse(v.begin(), v.end());
        assert(ret == v);
    }

    cout << "All testcases passed" << endl;

}

void benchMark()
{
    unique_ptr<BaseSolution> pSolution = make_unique<Solution>();
    unique_ptr<BaseSolution> pSolution1 = make_unique<Solution1>();
    unique_ptr<BaseSolution> pSolution2 = make_unique<Solution2>();

    test(move(pSolution));
    test(move(pSolution1));
    test(move(pSolution2));
}


int main(int argc, char **argv)
{
    benchMark();
    
    return 0;
}

File: src/6_printListFromTailToHead.py
import offerhelper
from offerhelper import ListNode

class Solution:
    """Simple use a stack."""
    def printListFromTailToHead(self, listNode):
        stack = []
        while listNode:
            stack.append(listNode.val)
            listNode = listNode.next
        return stack[::-1]


class Solution1:
    """Insert at the begining."""
    def printListFromTailToHead(self, listNode):
        head = ListNode(0)
        while listNode:
            saved_next = listNode.next
            listNode.next = head.next
            head.next = listNode
            listNode = saved_next

        li = []
        listNode = head.next
        while listNode:
            li.append(listNode.val)
            listNode = listNode.next

        return li


class Solution2:
    """Recursion solution."""
    # when size of listNode is larger than 995,
    # maximum recursion error will be raised
    def helper(self, listNode, li):
        if not listNode:
            return
        self.helper(listNode.next, li)
        li.append(listNode.val)

    def printListFromTailToHead(self, listNode):
        li = []
        self.helper(listNode, li)
        return li


class Solution3:
    """Recursion solution."""
    # when size of listNode is larger than 995,
    # maximum recursion error will be raised
    def printListFromTailToHead(self, listNode):
        li = []
        if listNode:
            li.extend(self.printListFromTailToHead(listNode.next))
            li.append(listNode.val)
        return li


def test(solution):
    testcases = (
            [],  # empty linked list
            [1],  # only one node
            [1, 2, 3],  # simple list
            [1, 2, 2, 3],  # simple list with duplicated values
            offerhelper.gen_random_list(995, 0, 10000)  # large random linked list
            )

    for testcase in testcases:
        head = offerhelper.create_linked_list(testcase)
        assert solution.printListFromTailToHead(head) == testcase[::-1]


def bench_mark():
    solution = Solution()
    solution1 = Solution1()
    solution2 = Solution2()
    solution3 = Solution3()

    # head = offerhelper.create_linked_list([1, 2, 3])
    # solution1.printListFromTailToHead(head)

    test(solution)
    test(solution)
    test(solution)
    test(solution)

    print("All testcases pass")


if __name__ == '__main__':
    bench_mark()

